### PR TITLE
Enable Alibaba Cloud provider instances with expected ProviderID

### PR DESCRIPTION
**Pull request by kwoodson**
**- What I did**
This code enables the Alibaba Cloud provider platform on OpenShift.  The Alibaba provider requires that the instances have a specific format in which the [documentation](https://github.com/kubernetes/cloud-provider-alibaba-cloud/blob/master/docs/getting-started.md) states should be in `"${REGION}.${INSTANCE_ID}"`.

This patch follows the pattern for other providers which sets an environment variable named `KUBELET_NODE_NAME` in `/etc/systemd/system/kubelet.service.d/`. The environment variable is set through creating a service that upon start will query the metadata service for the region and the instanceid.

This pull request requires #2772.

**- How to verify it**
I was able to verify that this code works by deploying the following work:
- https://github.com/openshift/installer/pull/5018 with go.mod dependencies from https://github.com/openshift/installer/pull/5216
- Creating a custom build of the cluster-cloud-controller-manager-operator from Alibaba: https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/119
- Creating a custom build of the CCM from https://github.com/openshift/cloud-provider-alibaba-cloud with a specific fix for https://github.com/openshift/cloud-provider-alibaba-cloud/issues/10

**- Description for the changelog**
Enabling support for Alibaba Cloud provider node names

### CODE REVIEW

**Comment by sinnykumari** [comment]
There is already an open PR for updating OpenShift libs https://github.com/openshift/machine-config-operator/pull/2772. Can you please consolidate changes into one PR and close one of them?

**Comment by kwoodson** [comment]
@sinnykumari Thanks for looking. There is some discussion going on for #2772 currently regarding the `PodSecurity: True`. 

I feel better about leaving them separate as the 2772 updates the API, library-go, and client-go which is independent of these changes. If you would prefer to join them together I can do so.

**Comment by sinnykumari** [comment]
Even this PR is updating `PodSecurity: True`  in the kubelet config https://github.com/openshift/machine-config-operator/pull/2777/files#diff-7e37987828daa82184819a87de503b6546be9a83519ff94f416447e4b3814d11 .

**Comment by kwoodson** [comment]
@sinnykumari This is built on top of 2772 so it includes the library updates.

**Review by kikisdeliveryservice — DISMISSED** [review]
Can you update the OWNERS file for alibaba:
https://github.com/openshift/machine-config-operator/blob/master/OWNERS_ALIASES

And invoke it in your new dir like baremetal & co do: https://github.com/openshift/machine-config-operator/tree/master/templates/common/baremetal

**Comment by bd233** [comment]
@kwoodson 
Hello Kenny. Regarding the issue of 'ProviderID', I would like to know whether I need to make some changes in the Installer repo in addition to the modification of this library？

**Comment by kikisdeliveryservice** [comment]
just adding a hold until https://github.com/openshift/machine-config-operator/pull/2772 merges

/hold

**Comment by sinnykumari** [comment]
Removing hold now since #2772 has got merged
/hold cancel

**Comment by kikisdeliveryservice** [comment]
@rvanderp3 @fabianofranz PTAL

**Review by kikisdeliveryservice — COMMENTED** [review]
I'm a little (a lot) rusty on ratcheting in a new cloud provider, but just to understand this cloud not using cloud.conf flag?
https://github.com/openshift/machine-config-operator/blob/4f555d05550bca9a67f5354afa2721cea5e36b02/pkg/controller/template/render.go#L335
(see 335-396)

I'd expect that this Test should also be updated: https://github.com/openshift/machine-config-operator/blob/4f555d05550bca9a67f5354afa2721cea5e36b02/pkg/controller/template/render_test.go#L26

**Comment by kwoodson** [comment]
> I'm a little (a lot) rusty on ratcheting in a new cloud provider, but just to understand this cloud not using cloud.conf flag?
> 
> https://github.com/openshift/machine-config-operator/blob/4f555d05550bca9a67f5354afa2721cea5e36b02/pkg/controller/template/render.go#L335
> 
> 
> (see 335-396)
> I'd expect that this Test should also be updated:
> 
> https://github.com/openshift/machine-config-operator/blob/4f555d05550bca9a67f5354afa2721cea5e36b02/pkg/controller/template/render_test.go#L26

Thanks @kikisdeliveryservice. I've been out for a couple of days. 

Regarding the `cloudProvider` function this is handled by the `isCloudProviderExternal` which came in through the `library-go` update in #2772. The out-of-tree provider should return `external` as its value and is set in the `kubelet.service` after the mco runs.

I have added a test for Alibaba that should return `external`. Thanks for this one I had missed it!

**Review by sinnykumari — APPROVED** [review]
LGTM

**Comment by kikisdeliveryservice** [comment]
waiting for final review, since you're new owners of this can you PTAL @rvanderp3 @fabianofranz?

File: manifests/controllerconfig.crd.yaml
Comment (kikisdeliveryservice) [reply manifests/controllerconfig.crd.yaml:265]: Also is this camelcase on purpose? bc the file names don't have it and seems prone to errors?

ibmcloud don't camel

**Review by kikisdeliveryservice — COMMENTED** [review]
Perusing a bit before I log of to get a jump for next week: 

I think there is a test to update: https://github.com/openshift/machine-config-operator/blob/8d2a734457e4e6ab17e3d14ee909edf0a95c77ee/pkg/controller/template/render_test.go#L284

and adding a file here:

https://github.com/openshift/machine-config-operator/tree/master/pkg/controller/template/test_data

I didn't notice this in my review before apologies.

**Comment by kikisdeliveryservice** [comment]
Upgrade failure seems unrelated but is quite red lately so let's retest:

/test e2e-agnostic-upgrade

**Comment by kwoodson** [comment]
@kikisdeliveryservice Thanks, I'll update that test on Monday morning as well as get my reviewers to TAL. Your reviews are much appreciated!

**Comment by JoelSpeed** [comment]
If it comes in useful, I rebased my crd-gen branch on top of today's master branch, and updated the generated CRD based on that, https://github.com/JoelSpeed/machine-config-operator/tree/crd-gen-2021-10-11, this should show you what CRD gen would be doing in this case

**Comment by kwoodson** [comment]
Looks like a busted test:
```
1/62 Tests Failed
...
Failed while waiting on imagestream import
{"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-10-11T21:52:10Z"}
error: failed to execute wrapped command: exit status 1
```

**Comment by kwoodson** [comment]
@kikisdeliveryservice I had a team member take a look at this. Apologies for the delay we have been in training the previous week. Did you have anything else required for this PR?

**Pull request merged**
